tests/end-to-end/upgrade/pkg/tests/service-catalog/app_broker.go

```diff
@@ -0,0 +1,404 @@
+	apiServiceId        = "api-service-id"
+	eventsServiceId     = "events-service-id"
+	apiBindingName      = "app-binding"
+	apiBindingUsageName = "app-binding"
+	gatewayUrl          = "https://gateway.local"
+)
+
+type AppBrokerUpgradeTest struct {
+	ServiceCatalogInterface clientset.Interface
+	K8sInterface            kubernetes.Interface
+	BUInterface             bu.Interface
+	AppBrokerInterface      appBroker.Interface
+	AppConnectorInterface   appConnector.Interface
+}
+
+func NewAppBrokerUpgradeTest(scCli clientset.Interface,
+	k8sCli kubernetes.Interface,
+	buCli bu.Interface,
+	abCli appBroker.Interface,
+	acCli appConnector.Interface) *AppBrokerUpgradeTest {
+	return &AppBrokerUpgradeTest{
+		ServiceCatalogInterface: scCli,
+		K8sInterface:            k8sCli,
+		BUInterface:             buCli,
+		AppBrokerInterface:      abCli,
+		AppConnectorInterface:   acCli,
+	}
+}
+
+type appBrokerFlow struct {
+	baseFlow
+
+	scInterface           clientset.Interface
+	k8sInterface          kubernetes.Interface
+	buInterface           bu.Interface
+	appBrokerInterface    appBroker.Interface
+	appConnectorInterface appConnector.Interface
+}
+
+func (ut *AppBrokerUpgradeTest) CreateResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
+	return ut.newFlow(stop, log, namespace).CreateResources()
+}
+
+func (ut *AppBrokerUpgradeTest) TestResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
+	return ut.newFlow(stop, log, namespace).TestResources()
+}
+
+func (ut *AppBrokerUpgradeTest) newFlow(stop <-chan struct{}, log logrus.FieldLogger, namespace string) *appBrokerFlow {
+	return &appBrokerFlow{
+		baseFlow: baseFlow{
+			log:       log,
+			stop:      stop,
+			namespace: namespace,
+
+			buInterface:  ut.BUInterface,
+			scInterface:  ut.ServiceCatalogInterface,
+			k8sInterface: ut.K8sInterface,
+		},
+
+		appBrokerInterface:    ut.AppBrokerInterface,
+		appConnectorInterface: ut.AppConnectorInterface,
+		scInterface:           ut.ServiceCatalogInterface,
+	}
+}
+
+func (f *appBrokerFlow) CreateResources() error {
+	// iterate over steps
+	for _, fn := range []func() error{
+		f.createApplication,
+		f.createApplicationMapping,
+		f.deployEnvTester,
+		f.waitForEnvTester,
+		f.waitForClassAndPlans,
+		f.createAPIServiceInstance,
+		f.createEventsServiceInstance,
+		f.waitForInstances,
+		f.createAPIServiceBinding,
+		f.createAPIServiceBindingUsage,
+	} {
+		err := fn()
+		if err != nil {
+			f.logReport()
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *appBrokerFlow) TestResources() error {
+	// iterate over steps
+	for _, fn := range []func() error{
+		f.verifyEventActivation,
+		f.verifyEnvTesterHasGatewayInjected,
+		f.deleteAPIServiceBindingUsage,
+		f.deleteAPIServiceBinding,
+		f.deleteAPIServiceInstance,
+		f.deleteEventsServiceInstance,
+		f.waitForInstancesDeleted,
+		f.deleteApplicationMapping,
+		f.deleteApplication,
+		f.waitForClassesRemoved,
```

Comment:
shouldn't we also check that after deleting the SBU the example deploy does not have injected environment variable?

In that way, we will check also that unbinding from our business perspective working as expected. E.g. we had a problem that when we removed the SBU for kubeless function then the underlying pod was not restarted and still have injected both credentials and URL. New assertion in that test will detect that unexpected behavior.

Reply:
new assertion added